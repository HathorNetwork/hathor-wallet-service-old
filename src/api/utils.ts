/**
 * Copyright (c) Hathor Labs and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { APIGatewayProxyResult } from 'aws-lambda';
import { ServerlessMysql } from 'serverless-mysql';

import { ApiError } from '@src/api/errors';
import { StringMap } from '@src/types';
import { closeDbConnection } from '@src/utils';

/**
 * Close database connection and get error object.
 *
 * @param mysql - The database connection
 * @param error - ApiError return code
 * @param extra - Extra data to be sent on the body of the error object
 * @returns The error object
 */
export const closeDbAndGetError = async (
  mysql: ServerlessMysql,
  error: ApiError,
  extra?: StringMap<unknown>,
): Promise<APIGatewayProxyResult> => {
  await closeDbConnection(mysql);
  const body = { success: false, error, ...extra };
  return {
    statusCode: 200,
    body: JSON.stringify(body),
  };
};
