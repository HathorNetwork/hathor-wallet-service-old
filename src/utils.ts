/**
 * Copyright (c) Hathor Labs and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { createHash, HexBase64Latin1Encoding } from 'crypto';
import serverlessMysql, { ServerlessMysql } from 'serverless-mysql';
import hathorLib from '@hathor/wallet-lib';
/**
 * Calculate the double sha256 hash of the data.
 *
 * @remarks
 * If encoding is provided a string will be returned; otherwise a Buffer is returned.
 *
 * @param data - Data to be hashed
 * @param encoding - The encoding of the returned object
 * @returns The sha256d hash of the data
 */
export const sha256d = (data: string, encoding: HexBase64Latin1Encoding): string => {
  const hash1 = createHash('sha256');
  hash1.update(data);
  const hash2 = createHash('sha256');
  hash2.update(hash1.digest());
  return hash2.digest(encoding);
};

/**
 * Get the wallet id given the xpubkey.
 *
 * @param xpubkey - The xpubkey
 * @returns The wallet id
 */
export const getWalletId = (xpubkey: string): string => (
  sha256d(xpubkey, 'hex')
);

/**
 * Get the current Unix timestamp, in seconds.
 *
 * @returns The current Unix timestamp in seconds
 */
export const getUnixTimestamp = (): number => (
  Math.round((new Date()).getTime() / 1000)
);

/**
 * Get a database connection.
 *
 * @returns The database connection
 */
export const getDbConnection = (): ServerlessMysql => (
  serverlessMysql({
    config: {
      host: process.env.DB_ENDPOINT,
      database: process.env.DB_NAME,
      user: process.env.DB_USER,
      // TODO if not on local env, get IAM token
      // https://aws.amazon.com/blogs/database/iam-role-based-authentication-to-amazon-aurora-from-serverless-applications/
      password: process.env.DB_PASS,
    },
  })
);

export const closeDbConnection = async (mysql: ServerlessMysql): Promise<void> => {
  if (process.env.STAGE === 'local') {
    // mysql.end() leaves the function hanging in the local environment. Some issues:
    // https://github.com/jeremydaly/serverless-mysql/issues/61
    // https://github.com/jeremydaly/serverless-mysql/issues/79
    //
    // It seems that's the expected behavior for local environment:
    // https://github.com/serverless/serverless/issues/470#issuecomment-205372006
    await mysql.quit();
  } else {
    await mysql.end();
  }
};

export const isAuthority = (tokenData: number): boolean => (
  (tokenData & hathorLib.constants.TOKEN_AUTHORITY_MASK) > 0    // eslint-disable-line no-bitwise
);
