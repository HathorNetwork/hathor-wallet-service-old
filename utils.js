import { createHash } from 'crypto';
import { Address, HDPublicKey, Networks } from 'bitcore-lib';
import serverlessMysql from 'serverless-mysql';

// TODO get from hathor-lib or maybe env?
const mainnet = Networks.add({
  name: 'mainnet',
  alias: 'production',
  pubkeyhash: 0x28,
  privatekey: 0x80,
  scripthash: 0x64,
  xpubkey: 0x0488b21e,
  xprivkey: 0x0488ade4,
  networkMagic: 0xf9beb4d9,
  port: 8333,
  dnsSeeds: []
});


export const getWalletId = (xpubkey) => {
  return sha256d(xpubkey, 'hex');
};

/*
 * input: <string> | <Buffer> | <TypedArray> | <DataView>
 */
export const sha256d = (input, encoding) => {
  const hash1 = createHash('sha256');
  hash1.update(input);
  const hash2 = createHash('sha256');
  hash2.update(hash1.digest());
  return hash2.digest(encoding);
};

export const getHathorAddresses = (xpubkey, startIndex, quantity) => {
  const addrMap = {};
  const xpub = HDPublicKey(xpubkey);
  for (let index = startIndex; index < startIndex + quantity; index++) {
    const key = xpub.derive(index);
    const address = Address(key.publicKey, mainnet);
    addrMap[address.toString()] = index;
  }
  return addrMap;
};

export const getUnixTimestamp = () => {
  return Math.round((new Date()).getTime() / 1000);
};

export const getDbConnection = () => {
  return serverlessMysql({
    config: {
      host: process.env.DB_ENDPOINT,
      database: process.env.DB_NAME,
      user: process.env.DB_USER,
      //TODO if not on local stage, get IAM token
      //https://aws.amazon.com/blogs/database/iam-role-based-authentication-to-amazon-aurora-from-serverless-applications/
      password: process.env.DB_PASS,
    }
  });
};
