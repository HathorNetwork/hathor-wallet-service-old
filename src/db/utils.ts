/**
 * Copyright (c) Hathor Labs and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable max-classes-per-file */
import { ServerlessMysql } from 'serverless-mysql';
import { getWalletId } from '@src/utils';
import {
  WalletStatus,
  Wallet,
  Tx,
  DbSelectResult,
  TokenBalanceMap,
  BalanceValue,
  WalletBalanceResult,
  StringMap,
  WalletBalance,
} from "@src/types";

/**
 * Begins a transaction on the current connection
 *
 * @param mysql - Database connection
 */
export const beginTransaction = async (
  mysql: ServerlessMysql,
): Promise<void> => {
  await mysql.query('START TRANSACTION');
};

/**
 * Commits the transaction opened on the current connection
 *
 * @param mysql - Database connection
 */
export const commitTransaction = async (
  mysql: ServerlessMysql,
): Promise<void> => {
  await mysql.query('COMMIT');
};

/**
 * Rollback the transaction opened on the current connection
 *
 * @param mysql - Database connection
 */
export const rollbackTransaction = async (
  mysql: ServerlessMysql,
): Promise<void> => {
  await mysql.query('ROLLBACK');
};

/* eslint-disable-next-line  @typescript-eslint/ban-types */
export async function transactionDecorator(_mysql: ServerlessMysql, wrapped: Function): Promise<Function> {
  return async function wrapper(...args) {
    try {
      await beginTransaction(_mysql);
      await wrapped.apply(this, args);
      await commitTransaction(_mysql);
    } catch (e) {
      await rollbackTransaction(_mysql);

      // propagate the error
      throw e;
    }
  };
}

/**
 * Returns a Wallet object from a db result row
 *
 * @param result - The result row to map to a Wallet object
 */
export const getWalletFromDbEntry = (entry: Record<string, unknown>): Wallet => ({
  walletId: getWalletId(entry.xpubkey as string),
  xpubkey: entry.xpubkey as string,
  authXpubkey: entry.auth_xpubkey as string,
  status: entry.status as WalletStatus,
  retryCount: entry.retry_count as number,
  maxGap: entry.max_gap as number,
  createdAt: entry.created_at as number,
  readyAt: entry.ready_at as number,
});

/**
 * Receive a DbSelectResult with multiple records and transform it in an array of Tx
 *
 * @param results
 * @returns Txs converted from DbSelectResult
 */
export const getTxsFromDBResult = (results: DbSelectResult): Tx[] => {
  const transactions = [];

  for (const result of results) {
    const tx: Tx = _mapTxRecord2Tx(result);

    transactions.push(tx);
  }

  return transactions;
};

/**
 * Receive a DbSelectResult with one record and transform it in a Tx
 *
 * @param results
 * @returns Tx converted from DbSelectResult
 */
export const getTxFromDBResult = (result: DbSelectResult): Tx => {
  const { 0: row } = result;
  return _mapTxRecord2Tx(row);
};

const _mapTxRecord2Tx = (record: Record<string, unknown>): Tx => (
  {
    txId: record.tx_id as string,
    timestamp: record.timestamp as number,
    version: record.version as number,
    voided: record.voided === 1,
    height: record.height as number,
    weight: record.weight as number,
  }
);

export class FromTokenBalanceMapToBalanceValueList {
  static convert(tokenBalanceMap: TokenBalanceMap): BalanceValue[] {
    const entryBalances = Object.entries(tokenBalanceMap.map);
    const balances = entryBalances.map((entry) => ({
      tokenId: entry[0],
      lockedAmount: entry[1].lockedAmount,
      lockedAuthorities: entry[1].lockedAuthorities.toJSON(),
      lockExpires: entry[1].lockExpires,
      unlockedAmount: entry[1].unlockedAmount,
      unlockedAuthorities: entry[1].unlockedAuthorities.toJSON(),
      totalAmountSent: entry[1].totalAmountSent,
      total: entry[1].total(),
    } as BalanceValue));
    return balances;
  }
}

export const sortBalanceValueByAbsTotal = (balanceA: BalanceValue, balanceB: BalanceValue): number => {
  if (Math.abs(balanceA.total) - Math.abs(balanceB.total) >= 0) return -1;
  return 0;
};

export class FromWalletBalanceMapToWalletBalanceResultList {
  static convert(walletBalanceMap: StringMap<WalletBalance>): WalletBalanceResult[] {
    const wallets: WalletBalanceResult[] = [];
    for (const wallet of Object.values(walletBalanceMap)) {
      // Sort by the tokens with the most balance
      const sortedBalanceList = FromTokenBalanceMapToBalanceValueList
        .convert(wallet.walletBalanceForTx)
        .sort(sortBalanceValueByAbsTotal);
      wallets.push({
        addresses: wallet.addresses,
        txId: wallet.txId,
        walletId: wallet.walletId,
        walletBalanceForTx: sortedBalanceList,
      });
    }

    return wallets;
  }
}
