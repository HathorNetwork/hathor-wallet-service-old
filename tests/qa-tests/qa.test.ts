import { config, HathorWalletServiceWallet, walletApi } from '@hathor/wallet-lib';
import { TEST_SEED } from '@tests/utils';

let wallet;


beforeAll(async () => {
  config.setNetwork(process.env.NETWORK);
  config.setWalletServiceBaseUrl('https://staging.wallet-service.hathor.network');
  config.setWalletServiceBaseWsUrl('https://ws.staging.wallet-service.hathor.network');

  wallet = new HathorWalletServiceWallet({
    requestPassword: () => {},
    seed: TEST_SEED,
    network: config.getNetwork() 
  });

  // Start the wallet like this will indirectly test the following endpoints:
  // - `/wallet/init`: by calling the walletApi.createWallet method
  // - `/wallet/status`: by calling the wallet.pollForWalletStatus method
  await wallet.start({ pinCode: '1234', password: '' });
}, 20000);

afterAll(async () => {
  await wallet.stop();
}, 20000);

test('GET /wallet/transactions/{txId}', async () => {
  const response = await wallet.getTxById("000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002");

  expect(response.success).toStrictEqual(true);
  expect(response.txTokens).toHaveLength(1);
  expect(response.txTokens[0]).toEqual({
      balance: -1,
      timestamp: 1681409837,
      tokenId: '00',
      tokenName: 'Hathor',
      tokenSymbol: 'HTR',
      txId: '000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002',
      version: 1,
      voided: false,
      weight: 16.8184
  });
}, 30000);

test('GET /wallet/proxy/transactions/{txId}', async () => {
  const response = await wallet.getFullTxById("000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d");

  expect(response).toMatchSnapshot();
}, 30000);

test('GET /wallet/proxy/transactions/{txId}/confirmation_data', async () => {
  const response = await wallet.getTxConfirmationData("000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d");

  expect(response).toEqual({
    "accumulated_bigger": true,
    "accumulated_weight": 70.12392675219921,
    "confirmation_level": 1,
    "stop_value": 69.90262876589468,
    "success": true,
  });
}, 30000);

test('GET /wallet/proxy/graphviz/neighbours', async () => {
  const response = await wallet.graphvizNeighborsQuery(
    "000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d",
    "verification",
    5
  );

  expect(response).toMatchSnapshot();
}, 30000);

test('GET /wallet/history', async () => {
  const response = await wallet.getTxHistory();

  // We expect 2 transactions to exist for our test wallet.
  // If someone uses the seed to do more transactions in the mainnet, this test will fail
  expect(response).toMatchSnapshot();
}, 30000);

test('GET /wallet/addresses', async () => {
  const addresses = [];

  for await (const a of wallet.getAllAddresses()) {
    addresses.push(a);
  }

  expect(addresses).toMatchSnapshot();
}, 30000);

test('POST /wallet/addresses/check_mine', async () => {
  const response = await wallet.checkAddressesMine([
    "HBCQgVR8Xsyv1BLDjf9NJPK1Hwg4rKUh62",
    "HPDWdurEygcubNMUUnTDUAzngrSXFaqGQc",
    "HLfGaQoxssGbZ4h9wbLyiCafdE8kPm6Fo4",
    "HJbeBkDqBewxkAqsnvvV2Ge8qu31Tmc3XK",
    "H9A6fG4JxT1HCktiaZwLQCCgsgUM7axndb"
  ]);

  expect(response).toMatchSnapshot();
}, 30000);

test.todo('GET /wallet/addresses/new');
test.todo('GET /wallet/utxos');
test.todo('GET /wallet/tx_outputs');
test.todo('GET /wallet/balances');
test.todo('GET /wallet/tokens');
test.todo('GET /wallet/tokens/{token_id}/details');


test.todo('/wallet/push/register');
test.todo('/wallet/push/update');
test.todo('/wallet/push/unregister/{deviceId}');

test.todo('POST /tx/proposal');
test.todo('PUT /tx/proposal/{txProposalId}');
test.todo('DELETE /tx/proposal/{txProposalId}');


test.todo('PUT /wallet/auth');

test.todo('GET /version');

test.todo('GET /metrics');

test.todo('/auth/token'); // TODO: I think this is indirectly tested when starting the wallet



// TODO: Lambdas that are not API endpoints:
// sendNotificationToDevice
// txPushRequested
// getLatestBlock
// onNewTxRequest
// onMinersListRequest
// onTotalSupplyRequest
// onHandleReorgRequest
// onNewTxEvent
// onNewNftEvent

// TODO: Lambdas that are CRONs:
// deleteStalePushDevices
// cleanUnsentTxProposalsUtxos
// onHandleOldVoidedTxs


// TODO: Websocket Lambdas
// wsConnect
// wsJoin
// wsTxNotifyNew
// wsTxNotifyUpdate
// wsAdminBroadcast
// wsAdminDisconnect
// wsAdminMulticast
