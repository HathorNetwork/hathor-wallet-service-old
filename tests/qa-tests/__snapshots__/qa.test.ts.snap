// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /wallet/addresses 1`] = `
Array [
  Object {
    "address": "HBCQgVR8Xsyv1BLDjf9NJPK1Hwg4rKUh62",
    "index": 0,
    "transactions": 2,
  },
  Object {
    "address": "HPDWdurEygcubNMUUnTDUAzngrSXFaqGQc",
    "index": 1,
    "transactions": 0,
  },
  Object {
    "address": "HEYCNNZZYrimD97AtoRcgcNFzyxtkgtt9Q",
    "index": 2,
    "transactions": 0,
  },
  Object {
    "address": "HPTtSRrDd4ekU4ZQ2jnSLYayL8hiToE5D4",
    "index": 3,
    "transactions": 0,
  },
  Object {
    "address": "HTYymKpjyXnz4ssEAnywtwnXnfneZH1Dbh",
    "index": 4,
    "transactions": 0,
  },
  Object {
    "address": "HUp754aDZ7yKndw2JchXEiMvgzKuXasUmF",
    "index": 5,
    "transactions": 0,
  },
  Object {
    "address": "HLfGaQoxssGbZ4h9wbLyiCafdE8kPm6Fo4",
    "index": 6,
    "transactions": 0,
  },
  Object {
    "address": "HV3ox5B1Dai6Jp5EhV8DvUiucc1z3WJHjL",
    "index": 7,
    "transactions": 0,
  },
  Object {
    "address": "HNWxs2bxgYtzfCpU6cJMGLgmqv7eGupTHr",
    "index": 8,
    "transactions": 0,
  },
  Object {
    "address": "H9Ef7qteC4vAoVUYx5mvP9jCfmZgU9rSvL",
    "index": 9,
    "transactions": 0,
  },
  Object {
    "address": "H7hxR75zsPzwfPWbrdkkFbKN2SiL2Lvyuw",
    "index": 10,
    "transactions": 0,
  },
  Object {
    "address": "HVCa4QJbHB6pkqvNkmQZD2vpmwTYMNdzVo",
    "index": 11,
    "transactions": 0,
  },
  Object {
    "address": "HBchgf1JLxwJzUg6epckK3YJn6Bq8XJMPV",
    "index": 12,
    "transactions": 0,
  },
  Object {
    "address": "HVWf61fwoj9Dx15NvWicqXQgGMYVYedSx4",
    "index": 13,
    "transactions": 0,
  },
  Object {
    "address": "H7PfxBmaqjoBisFRzpizoB9JcYSvoo8D2j",
    "index": 14,
    "transactions": 0,
  },
  Object {
    "address": "HC1NXVzGcVAd84QMfFngHiKyK2K8SUiTaL",
    "index": 15,
    "transactions": 0,
  },
  Object {
    "address": "HCqsSDrbs1cfqnF6QMUQkdGYXjEMyt9N3Y",
    "index": 16,
    "transactions": 0,
  },
  Object {
    "address": "H91QTkhcrFgrT3TAu1sgvxfDMqXj2TeF9K",
    "index": 17,
    "transactions": 0,
  },
  Object {
    "address": "HDVb7QR3MQ7Ypzs17PUBbvjdoLnPwq1nWU",
    "index": 18,
    "transactions": 0,
  },
  Object {
    "address": "H82H4FeDc23rKAxahcNBA3kNRjZAqRviEG",
    "index": 19,
    "transactions": 0,
  },
  Object {
    "address": "HULUsaW2h1zavJ1kXY8DewHVGcKdPfSYxk",
    "index": 20,
    "transactions": 0,
  },
]
`;

exports[`GET /wallet/history 1`] = `
Array [
  Object {
    "balance": -1,
    "timestamp": 1681409837,
    "txId": "000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002",
    "version": 1,
    "voided": 0,
  },
  Object {
    "balance": 1,
    "timestamp": 1681409807,
    "txId": "000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d",
    "version": 1,
    "voided": 0,
  },
]
`;

exports[`GET /wallet/proxy/graphviz/neighbours 1`] = `
"digraph {
	rankdir=RL
	node [shape=oval style=\\"\\"]
	\\"000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d\\" [label=\\"9c7d\\" penwidth=5.0 style=filled]
	\\"000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d\\" -> \\"0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a\\"
	\\"000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d\\" -> \\"000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4\\"
	\\"00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623\\" [label=3623]
	\\"00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623\\" -> \\"000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002\\"
	\\"00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623\\" -> \\"000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d\\"
	\\"000011dd2f550e18c7a60b2720b673a3fe36b3df510f3a39d4b7b08d067bef72\\" [label=ef72]
	\\"000011dd2f550e18c7a60b2720b673a3fe36b3df510f3a39d4b7b08d067bef72\\" -> \\"00000000454576da062712e48f970cf7d34d3ed9f6f7484339598e2e61eef48d\\"
	\\"000011dd2f550e18c7a60b2720b673a3fe36b3df510f3a39d4b7b08d067bef72\\" -> \\"00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623\\"
	\\"000001e40c51a86bbd56eb4a772c35d7a70cd471f2ffc8bbc3260253b05ec18a\\" [label=c18a]
	\\"000001e40c51a86bbd56eb4a772c35d7a70cd471f2ffc8bbc3260253b05ec18a\\" -> \\"0000025c2357235426fb38e75fe307bfa2fce398454f06e6585c0b87ae0d5dce\\"
	\\"000001e40c51a86bbd56eb4a772c35d7a70cd471f2ffc8bbc3260253b05ec18a\\" -> \\"000011dd2f550e18c7a60b2720b673a3fe36b3df510f3a39d4b7b08d067bef72\\"
	\\"00001c3b31932e29cbcfe3b31965a09315c7346b7ccb3eb1f7c29300aa0ed62f\\" [label=d62f]
	\\"00001c3b31932e29cbcfe3b31965a09315c7346b7ccb3eb1f7c29300aa0ed62f\\" -> \\"00004078dfca66d9130e28e05f6f5f3a3fad87ec7779e3cb955e498ab7e67bf4\\"
	\\"00001c3b31932e29cbcfe3b31965a09315c7346b7ccb3eb1f7c29300aa0ed62f\\" -> \\"000001e40c51a86bbd56eb4a772c35d7a70cd471f2ffc8bbc3260253b05ec18a\\"
	\\"00004078dfca66d9130e28e05f6f5f3a3fad87ec7779e3cb955e498ab7e67bf4\\" [label=\\"7bf4\\"]
	\\"00004078dfca66d9130e28e05f6f5f3a3fad87ec7779e3cb955e498ab7e67bf4\\" -> \\"000001e40c51a86bbd56eb4a772c35d7a70cd471f2ffc8bbc3260253b05ec18a\\"
	\\"00004078dfca66d9130e28e05f6f5f3a3fad87ec7779e3cb955e498ab7e67bf4\\" -> \\"0000025c2357235426fb38e75fe307bfa2fce398454f06e6585c0b87ae0d5dce\\"
	\\"0000025c2357235426fb38e75fe307bfa2fce398454f06e6585c0b87ae0d5dce\\" [label=\\"5dce\\"]
	\\"0000025c2357235426fb38e75fe307bfa2fce398454f06e6585c0b87ae0d5dce\\" -> \\"000011dd2f550e18c7a60b2720b673a3fe36b3df510f3a39d4b7b08d067bef72\\"
	\\"0000025c2357235426fb38e75fe307bfa2fce398454f06e6585c0b87ae0d5dce\\" -> \\"00000000454576da062712e48f970cf7d34d3ed9f6f7484339598e2e61eef48d\\"
	\\"00000000454576da062712e48f970cf7d34d3ed9f6f7484339598e2e61eef48d\\" [label=f48d]
	\\"00000000454576da062712e48f970cf7d34d3ed9f6f7484339598e2e61eef48d\\" -> \\"00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623\\"
	\\"00000000454576da062712e48f970cf7d34d3ed9f6f7484339598e2e61eef48d\\" -> \\"000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002\\"
	\\"000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002\\" [label=5002]
	\\"000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002\\" -> \\"000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d\\"
	\\"000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002\\" -> \\"0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a\\"
	\\"000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4\\" [label=\\"56d4\\"]
	\\"000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4\\" -> \\"00000446e4a3718d9e0fb92b0903648de6d71f1ec8fedf41266abb28b77a2ff5\\"
	\\"000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4\\" -> \\"0000061594cac2c50bcb9b948b53a272319e4d3ce217dc8c66e5c08ef9ecfc88\\"
	\\"0000061594cac2c50bcb9b948b53a272319e4d3ce217dc8c66e5c08ef9ecfc88\\" [label=fc88]
	\\"0000061594cac2c50bcb9b948b53a272319e4d3ce217dc8c66e5c08ef9ecfc88\\" -> \\"000007d8423f2b609f51d43002837ea99c8d992478f171c7d0cb36193c093aa3\\"
	\\"0000061594cac2c50bcb9b948b53a272319e4d3ce217dc8c66e5c08ef9ecfc88\\" -> \\"000005825591ea0e277b42f96f3b0b0aebc0cde19c6541dae2c9606e28c221ba\\"
	\\"000005825591ea0e277b42f96f3b0b0aebc0cde19c6541dae2c9606e28c221ba\\" [label=\\"21ba\\"]
	\\"000005825591ea0e277b42f96f3b0b0aebc0cde19c6541dae2c9606e28c221ba\\" -> \\"0000000082df841ba2c22be17788078d49f9cdc780a262bf3a9be3a1079d5983\\"
	\\"000005825591ea0e277b42f96f3b0b0aebc0cde19c6541dae2c9606e28c221ba\\" -> \\"000001a40317c9f9bfdfe6b292e347d04750354bcfe2145f345b3c31a7380373\\"
	\\"000001a40317c9f9bfdfe6b292e347d04750354bcfe2145f345b3c31a7380373\\" [label=0373]
	\\"0000000082df841ba2c22be17788078d49f9cdc780a262bf3a9be3a1079d5983\\" [label=5983]
	\\"0000000082df841ba2c22be17788078d49f9cdc780a262bf3a9be3a1079d5983\\" -> \\"000001a40317c9f9bfdfe6b292e347d04750354bcfe2145f345b3c31a7380373\\"
	\\"000007d8423f2b609f51d43002837ea99c8d992478f171c7d0cb36193c093aa3\\" [label=\\"3aa3\\"]
	\\"000007d8423f2b609f51d43002837ea99c8d992478f171c7d0cb36193c093aa3\\" -> \\"000005825591ea0e277b42f96f3b0b0aebc0cde19c6541dae2c9606e28c221ba\\"
	\\"000007d8423f2b609f51d43002837ea99c8d992478f171c7d0cb36193c093aa3\\" -> \\"0000000082df841ba2c22be17788078d49f9cdc780a262bf3a9be3a1079d5983\\"
	\\"00000446e4a3718d9e0fb92b0903648de6d71f1ec8fedf41266abb28b77a2ff5\\" [label=\\"2ff5\\"]
	\\"00000446e4a3718d9e0fb92b0903648de6d71f1ec8fedf41266abb28b77a2ff5\\" -> \\"0000061594cac2c50bcb9b948b53a272319e4d3ce217dc8c66e5c08ef9ecfc88\\"
	\\"00000446e4a3718d9e0fb92b0903648de6d71f1ec8fedf41266abb28b77a2ff5\\" -> \\"000007d8423f2b609f51d43002837ea99c8d992478f171c7d0cb36193c093aa3\\"
	\\"0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a\\" [label=\\"547a\\"]
	\\"0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a\\" -> \\"000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4\\"
	\\"0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a\\" -> \\"00000446e4a3718d9e0fb92b0903648de6d71f1ec8fedf41266abb28b77a2ff5\\"
}
"
`;

exports[`GET /wallet/proxy/transactions/{txId} 1`] = `
Object {
  "meta": Object {
    "accumulated_weight": 70.12392675219921,
    "children": Array [
      "00000000000000000a9664acb87f660c5893dcfd25b77e86d9082d68688a3e83",
      "000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002",
      "00000000000000000abed9e9d7537ac2839befbb62ab1a7d3e2efdb05d813727",
      "00001808469bd9cf6b3c5f26a02f5bc5e8a75e9dd7a222510145bb99f4b23623",
      "000000000000000007624d7a9cac9eb7938c7aa3a741ed3c694fd348d01c6e8d",
    ],
    "conflict_with": Array [],
    "first_block": "00000000000000000a9664acb87f660c5893dcfd25b77e86d9082d68688a3e83",
    "first_block_height": 3419474,
    "hash": "000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d",
    "height": 0,
    "min_height": 3419191,
    "received_by": Array [],
    "score": 0,
    "spent_outputs": Array [
      Array [
        0,
        Array [],
      ],
      Array [
        1,
        Array [
          "000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002",
        ],
      ],
    ],
    "twins": Array [],
    "validation": "full",
    "voided_by": Array [],
  },
  "spent_outputs": Object {
    "1": "000087bd0ed3cb0952b87bca3994b4f9cd014d60e6ddc82f82bdef3fb0a95002",
  },
  "success": true,
  "tx": Object {
    "hash": "000025451fcc127ef89cfa515b68b3b656b521fd4ff5ba66a81decfadd0b9c7d",
    "inputs": Array [
      Object {
        "decoded": Object {
          "address": "HVVHV7iTT4xDBZVRzw7FngGX4wqgu6YFpT",
          "timelock": null,
          "token_data": 0,
          "type": "P2PKH",
          "value": 800,
        },
        "index": 0,
        "script": "dqkU+/GoMUPeTDwiPQRXmV2WHts/Pl2IrA==",
        "token_data": 0,
        "tx_id": "00000000033c029cecb5d75c40e831e1e8e513dcad7679db2c73c87f8c9fc3d2",
        "value": 800,
      },
    ],
    "nonce": "284679",
    "outputs": Array [
      Object {
        "decoded": Object {
          "address": "HUWhe7WDNQXgarPrbepyqQax4JUHJHxQkv",
          "timelock": null,
          "token_data": 0,
          "type": "P2PKH",
          "value": 799,
        },
        "script": "dqkU8T4cYKfoSFmXZlERw/G7KKIJ+lKIrA==",
        "token_data": 0,
        "value": 799,
      },
      Object {
        "decoded": Object {
          "address": "HBCQgVR8Xsyv1BLDjf9NJPK1Hwg4rKUh62",
          "timelock": null,
          "token_data": 0,
          "type": "P2PKH",
          "value": 1,
        },
        "script": "dqkUM05UG96vCHbm/CDwZXBW4ELQNvSIrA==",
        "token_data": 0,
        "value": 1,
      },
    ],
    "parents": Array [
      "0000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a",
      "000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d4",
    ],
    "raw": "000100010200000000033c029cecb5d75c40e831e1e8e513dcad7679db2c73c87f8c9fc3d2000069463044022007ea6a0e39bb3739fe481d60be90d301c8f7a92eb09feef21c199c5bcb9f1e0702200270a71053763fee86bb98a2d491ca367e9beca5197cc0885240368fd57ac527210367ac671fe0ac779d4e8cd7beee99089b7b105891c21a7f7085b6030a9d1a804a0000031f00001976a914f13e1c60a7e8485997665111c3f1bb28a209fa5288ac0000000100001976a914334e541bdeaf0876e6fc20f0657056e042d036f488ac40316257c666df9a6438470f020000012259b65ef48e9dc09c0a22a6bdab4ca6735ee604cac9b519147d86547a000007d75b2da1a447e84409886ba8cb6cb1841319b4982383d66e5c864f56d400045807",
    "timestamp": 1681409807,
    "tokens": Array [],
    "version": 1,
    "weight": 17.384151840320193,
  },
}
`;

exports[`POST /wallet/addresses/check_mine 1`] = `
Object {
  "H9A6fG4JxT1HCktiaZwLQCCgsgUM7axndb": false,
  "HBCQgVR8Xsyv1BLDjf9NJPK1Hwg4rKUh62": true,
  "HJbeBkDqBewxkAqsnvvV2Ge8qu31Tmc3XK": false,
  "HLfGaQoxssGbZ4h9wbLyiCafdE8kPm6Fo4": true,
  "HPDWdurEygcubNMUUnTDUAzngrSXFaqGQc": true,
}
`;
